src/drivers/uart/uart.c

```diff
@@ -1 +1,101 @@
+#include "FreeRTOS.h"
+#include "semphr.h"
+
 #include "drivers/uart/uart.h"
+
+// Datatypes ----------------------------
+
+/**
+ * The data associated with a UART channel
+ */
+typedef struct
+{
+    UART_HandleTypeDef *handle; // HAL uart handle
+    SemaphoreHandle_t
+        transfer_complete;         // Communicate transfer complete between uart_write and the ISR
+    SemaphoreHandle_t write_mutex; // Allows tasks to line up to use uart_write
+} uart_handle_t;
+
+/**
+ * Maps each `uart_channel_t` to its `uart_handle_t` data
+ */
+uart_handle_t uart_channel_map[UART_CHANNEL_COUNT];
```

Comment:
Do we ever initialize this array? Consider that `uart_init()` expects the array to already be ready for writing.